File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap13/EventMethods.pl
#!/usr/bin/perl -w

### EventMethods.pl

use strict;
use Tk;

my $mw = MainWindow->new( -title => "Testing Callback" );            #(A)
my $b = $mw->Button( -text => "click here" )->pack( -side => 'top' );#(B)
my $c = $mw->Canvas()->pack( -side => 'bottom' );                    #(C)

$mw->bind( '<KeyPress>' => \&testEventMethods );                     #(D)
$b->bind( '<Button>' => \&testEventMethods );                        #(E)
$c->CanvasBind( '<Button>' => \&testEventMethods );                  #(F)

sub testEventMethods {                                               #(G)
    my $e = $Tk::event;                                              #(H)
    print "\n\nEvent object: $e\n";                                  #(I)
    print "Source widget: ", $e->W, "\n";                            #(J)
    print "Event object: ", $e->W->XEvent, "\n";                     #(K)
    print "Key (ASCII only): ", ($e->A ? $e->A : ''), "\n";          #(L)
    print "Key (any): ", ($e->K ? $e->K : ''), "\n";                 #(M)
    print "(x,y) coord relative to laptop screen: ", 
                                   $e->X, ", ", $e->Y, "\n";         #(N)
    print "(x,y) coord relative to widget: ", 
                                   $e->x, ", ", $e->y, "\n";         #(O)
}
MainLoop;                                                            #(P)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap13/EventMethods.py
#!/usr/bin/env python

### EventMethods.py

from Tkinter import *                                                #(A)

mw = Tk()                                                            #(B)
mw.title( "Testing Callback" )                                       #(C)       
mw.bind( '<KeyPress>', lambda e: testEventMethods() )                #(D)
b = Button( mw, text = "click here"  )                               #(E)
b.pack( side = 'top' )                                               #(F)
c = Canvas( mw )                                                     #(G)
c.pack( side = 'bottom' )                                            #(H)

mw.bind( '<KeyPress>', lambda e: testEventMethods( e ) )             #(I)
b.bind( '<Button>', lambda e: testEventMethods( e ) )                #(J)
c.bind( '<Button>', lambda e: testEventMethods( e ) )                #(K)

def testEventMethods( e ):                                           #(L)
    print "\n\nEvent object: %s" % e                                 #(M)
    print "Source widget (returned by e.widget): %s" % \
                                e.widget                             #(N)
    print "Source widget (returned by e.widget.winfo_name): %s" % \
                                e.widget.winfo_name()                #(O)
    if e.keysym_num != '??':                                         #(P)
        print "Key (ASCII only): %s" % e.keysym_num                  #(Q)
    else: print "Key (ASCII only):  "                                #(R)
    if e.keysym != '??':                                             #(S)
        print "Key (any): %s" % e.keysym                             #(T)
    else: print "Key (any): "                                        #(U)
    print "(x,y) coord relative to laptop screen: %s, %s" % \
                              (e.x_root, e.y_root)                   #(V)
    print "(x,y) coord relative to widget: %s, %s\n" %  (e.x, e.y)   #(W)

mainloop()                                                           #(X)
